src/hardware/sblaster.cpp

```diff
@@ -2894,28 +2849,84 @@ bool SB_GetAddress(uint16_t &sbaddr, uint8_t &sbirq, uint8_t &sbdma)
 	return (sbaddr != 0 && sbirq != 0 && sbdma != 0);
 }
 
-static void sblaster_callback(const uint32_t length)
+static void sblaster_mixer_callback([[maybe_unused]]const int requested_frames)
+{
+	// We can ignore requested frames as this function gets called in a loop until it gets what it needs
+	// Overflow is not a concern as extra frames will remain in the channel's buffer and get mixed on the next callback
+
+	// We must not block in fast-forward mode or the timings used by the mixer get confused
+	if (MIXER_InChipmunkMode() && soundblaster_mixer_queue.IsEmpty()) {
+		sb.chan->AddSilence();
+		return;
+	}
+	const auto frames = soundblaster_mixer_queue.Dequeue();
+	if (!frames) {
+		// Queue must be stopped, otherwise Dequeue() will block until some frames are available
+		sb.chan->AddSilence();
+		return;
+	}
+	// Weird double de-referencing syntax as the object type is std::optional<std::unique_ptr<AudioVector>>
+	(*frames)->AddSamples(sb.chan.get());
+}
+
+static void generate_frames(const int frames_requested)
 {
-	auto len = length;
+	static int ticks_of_silence = 0;
 
 	switch (sb.mode) {
 	case DspMode::None:
 	case DspMode::DmaPause:
-	case DspMode::DmaMasked: sb.chan->AddSilence(); break;
+	case DspMode::DmaMasked:
+	// Enqueue a tick's worth of silenced frames
+	// Some games (Tyrian for example) will switch to DmaMasked briefly (less than 5 ticks usually)
+	// We can't use AddSilence on the mixer thread as that asks for a blocksize of audio (usually over 10ms)
+	enqueue_frames(std::make_unique<AudioVectorM8S>(
+		frames_requested,
+		std::vector<int8_t>(frames_requested).data()
+	));
+
+	++ticks_of_silence;
+	if (ticks_of_silence > 5000) {
+		// Most games using DMA will run it constantly in that mode
+		// If we've hit 5 seconds of silence, we're probably not playing a DMA mode game
+		// Stop the mixer channel for performance reasons as this channel blocks waiting for the main thread to provide more audio
+		// We want this to be a conservative value as disabling the channel clears any pending audio
+
+		// This is different from the "sb.speaker_enable = false" state
+		// If the speaker is off, this function never gets called
+		// This is speaker on but playing silence
+		soundblaster_mixer_queue.Stop();
+		soundblaster_mixer_queue.Clear();
+		sb.chan->Enable(false);
+	}
+	break;
 
 	case DspMode::Dac:
-		// GenerateDACSound(len);
-		// break;
-		if (!sb.dac.used) {
-			sb.mode = DspMode::None;
-			return;
-		}
-
-		sb.chan->AddStretched(sb.dac.used, sb.dac.data);
+	// No-op if channel is already running
+	soundblaster_mixer_queue.Start();
+	sb.chan->Enable(true);
+	ticks_of_silence = 0;
+
+	if (sb.dac.used > 0) {
+		enqueue_frames(std::make_unique<AudioVectorStretched>(sb.dac.used, sb.dac.data));
```

Comment:
As @interloper98 has pointed out, we shouldn't be creating & destroying these buffers on the heap all the time. You could just extract them into temp vars and make them static, or turn them into instance variables, then they should auto-size themselves on demand.

Reply:
It would need to be an object pool.  Theoretically a game could switch audio formats from tick to tick so I'm enqueuing objects containing 1ms worth of audio each.  It's not ideal but it was the simplest I could think to do.

This isn't actually the thing @interloper98 measured though.  I'll comment what he was seeing down below.

Reply:
Uh, right, wasn't quite aware of that. That surely complicates things a lot, so it might be better to leave this as is if the performance impact is a non-issue. An object pool is a major complication so we should have a very good reason to introduce such a thing.

Reply:
There's some other strategies we could do.  Maybe converting the audio on the main thread before enqueuing it.  It would need quite a bit of refactoring though.  The convert function is meant to be called from `AddSamples()` which we can't do on the main thread.

I bench-marked with Quake which uses DMA sound running all the time and it didn't hurt over-all emulator performance compared to `main`.

Ideally, we can make DMA thread-safe at some point to make this a non-issue and, more importantly, further reduce audio stutters.

Reply:
Oh yeah, that's the end goal that all audio devices publish samples to the mixer as floats in the [-1.0, 1.0] range. Not for performance reasons, just it's conceptually a bit of a mess that the mixer does all those conversions IMO.

Also, only the SB deals with these different sample formats now AFAIK, so we don't even need global conversion helper functions; all of them could be moved into the SB code.

It's a bit too late for that, but if I knew you were going to do this, I would've suggested to do the float conversion first so the mixer has a single `AddSamples` method that deals with floats. Then you could have spared yourself the effort of templating your audio buffer thingy, which will eventually go away anyway.

Reply:
>  Also, only the SB deals with these different sample formats now AFAIK

There are some others.  Just right away I see CDROM image and Tandy DAC calling `AddSamples` with integer types.  The `AddSamples` function also handles mono to stereo and endianess conversion.

We'd still need all these conversion functions to exist somewhere.  Maybe we can de-couple it from the mixer at some point but probably not in this PR.

Converting to floating-point + stereo + native endianess is cheap so can be done in the main thread to simplify this SoundBlaster queue.

Resampling is much more expensive and should probably be done by the mixer.  Resampling is coupled to `AddSamples` at the moment.

Reply:
I agree with all that @weirddan455. Some lightweight non-templated conversion functions, the the mixer only has a single stereo add samples method that deals with normalised floats, and resampling in the mixer thread. But that's future PR material, yeah.